provider/db.go

```diff
@@ -0,0 +1,424 @@
+		err = rows.Scan(&c.ID, &c.RenterId, &c.ProviderId, &c.StorageSpace, &c.RenterSignature, &c.ProviderSignature, &startDate, &endDate)
+		if err != nil {
+			return nil, err
+		}
+		c.StartDate, _ = time.Parse(time.RFC3339, startDate)
+		c.EndDate, _ = time.Parse(time.RFC3339, endDate)
+		contracts = append(contracts, c)
+	}
+	return contracts, nil
+}
+
+func (p *Provider) GetAllContracts() ([]*core.Contract, error) {
+	rows, err := p.db.Query(`SELECT ContractId, RenterId, ProviderId, StorageSpace, 
+		RenterSignature, ProviderSignature, StartDate, EndDate FROM contracts`)
+	if err != nil {
+		return nil, err
+	}
+	var contracts []*core.Contract
+
+	for rows.Next() {
+		c := &core.Contract{}
+		// scan does not parse these directly into time.Time correctly
+		var startDate string
+		var endDate string
+		err = rows.Scan(&c.ID, &c.RenterId, &c.ProviderId, &c.StorageSpace, &c.RenterSignature, &c.ProviderSignature, &startDate, &endDate)
+		if err != nil {
+			return nil, err
+		}
+		c.StartDate, _ = time.Parse(time.RFC3339, startDate)
+		c.EndDate, _ = time.Parse(time.RFC3339, endDate)
+		contracts = append(contracts, c)
+	}
+	return contracts, nil
+}
+
+func (p *Provider) InsertBlock(renterId string, blockId string, size int64) error {
+	stmt, err := p.db.Prepare(`INSERT INTO blocks (RenterId, BlockId, Size) VALUES (?, ?, ?)`)
+	if err != nil {
+		return err
+	}
+	_, err = stmt.Exec(renterId, blockId, size)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Provider) DeleteBlockById(blockId string) error {
+	stmt, err := p.db.Prepare(`DELETE from blocks where BlockId=?`)
+	if err != nil {
+		return err
+	}
+	_, err = stmt.Exec(blockId)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Provider) GetBlocksByRenter(renterId string) ([]*BlockInfo, error) {
+
+	query := fmt.Sprintf(`SELECT BlockId, Size FROM blocks where RenterId='%s'`, renterId)
+	rows, err := p.db.Query(query)
+	if err != nil {
+		return nil, err
+	}
+	var blocks []*BlockInfo
+
+	for rows.Next() {
+		b := &BlockInfo{}
+		err = rows.Scan(&b.BlockId, &b.Size)
+		if err != nil {
+			return nil, err
+		}
+		blocks = append(blocks, b)
+	}
+	return blocks, nil
+}
+
+func (p *Provider) GetAllBlocks() ([]*BlockInfo, error) {
+
+	rows, err := p.db.Query(`SELECT RenterId, BlockId, Size FROM blocks`)
+	if err != nil {
+		return nil, err
+	}
+	var blocks []*BlockInfo
+
+	for rows.Next() {
+		b := &BlockInfo{}
+		err = rows.Scan(&b.RenterId, &b.BlockId, &b.Size)
+		if err != nil {
+			return nil, err
+		}
+		blocks = append(blocks, b)
+	}
+	return blocks, nil
+}
+
+//  Loads basic memory objects from db
+//  These will be recalculated based on db state at each restart
+//  (potentially useful for maintenance also)
+// - provider.StorageReserved
+// - provider.StorageUsed
+// - provider.TotalBlocks
+// - provider.TotalContracts
+// - provider.renters {
+// 	   - StorageUsed
+//     - StorageReserved
+//   }
+func (p *Provider) LoadDBIntoMemory() error {
```

Comment:
Agree.

Reply:
I think that recalculating these on startup actually makes these fields more resilient (as they can reconstruct from the elements in the DB on application restart if something happens to get out of sync). 

These fields are also incremented in memory in their respective methods.

Reply:
I don't see a real concern with keeping a few integer fields in sync with the DB, as long as they're incremented and decremented correctly. Another potential issue here is that this affects GUI start up time since it must wait for this work to be done before the provider server starts. I approve this for now, though.